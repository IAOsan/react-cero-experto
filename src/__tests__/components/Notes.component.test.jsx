import { vi } from 'vitest';
import '@testing-library/jest-dom';
import { renderWithProviders, setupUser } from '../test-utils';
import Notes from '../../components/Notes.component';
import {
	changeNote,
	addImageToNote,
	updateNote,
	removeNote,
} from '../../store/slices/notes/notes.slice';

vi.mock('../../store/slices/notes/notes.slice', async () => {
	const org = await vi.importActual('../../store/slices/notes/notes.slice');
	return {
		...org,
		changeNote: vi.fn(() => vi.fn()),
		addImageToNote: vi.fn(() => vi.fn()),
		updateNote: vi.fn(() => vi.fn()),
		removeNote: vi.fn(() => vi.fn()),
	};
});

const user = setupUser();
const activeNote = {
	title: 'title',
	createdAt: '2022-08-21T22:47:34.493Z',
	file: null,
	body: 'body',
	id: '7O03b76r5dQiK6XP5zBg',
};

describe('<Notes />', () => {
	beforeEach(() => {
		vi.clearAllMocks();
	});

	it('should be renders correctly', () => {
		const state = {
			entities: {
				notes: {
					active: activeNote,
				},
			},
		};
		const wrapper = renderNotes(state);

		expect(wrapper.container.firstElementChild).toMatchSnapshot();
		expect(wrapper.getByRole('navigation')).toBeInTheDocument();
		expect(
			wrapper.getByPlaceholderText(/some awesome title/i)
		).toBeInTheDocument();
		expect(
			wrapper.getByPlaceholderText(/what happend today/i)
		).toBeInTheDocument();
		expect(wrapper.queryByRole('img')).toBeNull();
		expect(
			wrapper.getByRole('button', { name: /remove note/i })
		).toBeInTheDocument();
	});

	it('should render note image if exists', () => {
		const imageUrl =
			'https://res.cloudinary.com/dflvdlaev/image/upload/v1661202277/zk2o6c2ubozleypnacvb.jpg';
		const state = {
			entities: {
				notes: {
					active: {
						...activeNote,
						imageUrl,
					},
				},
			},
		};
		const wrapper = renderNotes(state);

		const img = wrapper.getByRole('img');
		expect(img).toBeInTheDocument();
		expect(img.src).toBe(imageUrl);
	});

	it('should change active note', async () => {
		const state = {
			entities: {
				notes: {
					active: activeNote,
				},
			},
		};
		const wrapper = renderNotes(state);

		await user.type(
			wrapper.getByPlaceholderText(/some awesome title/i),
			' updated'
		);

		expect(changeNote).toHaveBeenLastCalledWith({
			title: activeNote.title + ' updated',
			body: activeNote.body,
		});
	});

	it('should add an image to active note', async () => {
		renderNotes();
		const file = new File([new ArrayBuffer(1)], 'file.jpg', {
			type: 'image/jpg',
		});

		await user.upload(document.querySelector('#file_selector'), file);

		expect(addImageToNote).toHaveBeenCalledWith(expect.any(File));
	});

	it('should save note', async () => {
		const wrapper = renderNotes();

		await user.click(wrapper.getByRole('button', { name: /save/i }));

		expect(updateNote).toHaveBeenCalled();
	});

	it('should remove active note', async () => {
		const state = {
			entities: {
				notes: {
					active: activeNote,
				},
			},
		};
		const wrapper = renderNotes(state);

		await user.click(wrapper.getByRole('button', { name: /remove note/i }));

		expect(removeNote).toHaveBeenCalledWith(activeNote.id);
	});
});

function renderNotes(state = {}) {
	return renderWithProviders(<Notes />, {
		preloadedState: state,
	});
}
