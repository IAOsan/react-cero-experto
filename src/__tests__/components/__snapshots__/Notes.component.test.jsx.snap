// Vitest Snapshot v1

exports[`<Notes /> > should be renders correctly 1`] = `
<section
  class="notes anima-fade-in-bck"
>
  <nav
    class="bg-primary shadow-22 text-light w-100 flex flex-ai-c flex-jc-sb notes__navbar"
  >
    <p>
      Sun Aug 21 2022
    </p>
    <input
      accept="image/*"
      id="file_selector"
      style="display: none;"
      type="file"
    />
    <div
      class="spacing-x-8"
    >
      <button
        class="btn btn--default btn--primary"
        type="button"
      >
        Picture
      </button>
      <button
        class="btn btn--default btn--primary"
        type="button"
      >
        Save
      </button>
    </div>
  </nav>
  <div
    class="flex notes__body"
  >
    <div
      class="form__group"
    >
      <input
        class="form__control form--notes__control"
        id="inputTitle"
        name="title"
        placeholder="Some awesome title"
        type="text"
        value="title"
      />
    </div>
    <div
      class="form__group h-100"
    >
      <textarea
        class="form__control form--notes__control form--notes__textarea h-100"
        id="inputBody"
        name="body"
        placeholder="What happend today"
      >
        body
      </textarea>
    </div>
    <hr
      class="my-20"
    />
    <button
      class="btn btn--default notes__remove-btn"
    >
      <svg
        class="icon"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M9 2a1 1 0 00-.894.553L7.382 4H4a1 1 0 000 2v10a2 2 0 002 2h8a2 2 0 002-2V6a1 1 0 100-2h-3.382l-.724-1.447A1 1 0 0011 2H9zM7 8a1 1 0 012 0v6a1 1 0 11-2 0V8zm5-1a1 1 0 00-1 1v6a1 1 0 102 0V8a1 1 0 00-1-1z"
          fill-rule="evenodd"
        />
      </svg>
      <span>
        Remove note
      </span>
    </button>
  </div>
</section>
`;
